python/ray/tests/test_autoscaler_drain_node_api.py

```diff
@@ -0,0 +1,77 @@
+import platform
+import time
+
+import pytest
+
+import ray
+from ray.cluster_utils import AutoscalingCluster
+import ray.ray_constants as ray_constants
+from ray.tests.test_autoscaler_fake_multinode import BASIC_CLUSTER_PARAMS
+from ray._private.test_utils import get_error_message, init_error_pubsub
+
+
+@pytest.mark.skipif(
+    platform.system() == "Windows", reason="Failing on Windows.")
+def test_drain_api(shutdown_only):
+    """E2E test of the autoscaler's use of the DrainNode API.
+
+    Adapted from test_autoscaler_fake_multinode.py.
+
+    The strategy is to mock out Ray node process termination in
+    FakeMultiNodeProvider, leaving node termination to the DrainNode API.
+
+    Scale-down is verified by `ray.cluster_resources`. It is verified that
+    no removed_node errors are issued adter scale-down.
+
+    Validity of this test depends on the current implementation of DrainNode.
+    DrainNode currently works by asking the GCS to de-register and shut down
+    Ray nodes.
+    """
+
+    cluster = AutoscalingCluster(
+        **BASIC_CLUSTER_PARAMS, _leave_termination_to_drain_api=True)
+    try:
+        cluster.start()
+        ray.init("auto")
+
+        # Triggers the addition of a GPU node.
+        @ray.remote(num_gpus=1)
+        def f():
+            print("gpu ok")
+
+        # Triggers the addition of a CPU node.
+        @ray.remote(num_cpus=3)
+        def g():
+            print("cpu ok")
+
+        ray.get(f.remote())
+        ray.get(g.remote())
+
+        # Verify scale-up
+        resources = ray.cluster_resources()
+        assert resources["CPU"] == 8
+        assert resources["GPU"] == 1
+        # Sleep for double the idle timeout of 6 seconds.
+        time.sleep(12)
+
+        # Verify scale-down
+        resources = ray.cluster_resources()
+        assert resources["CPU"] == 2
+        assert resources.get("GPU", 0) == 0
+
+        # Check that no errors were raised while draining nodes.
+        # (Logic copied from test_failure4::test_gcs_drain.)
+        try:
+            p = init_error_pubsub()
+            errors = get_error_message(
+                p, 1, ray_constants.REMOVED_NODE_ERROR, timeout=5)
+            assert len(errors) == 0
```

Comment:
@rkooo567 @ericl Does this block make sense?

Reply:
Yeah this looks good to me. You can probably confirm if it works by comparing before & after this PR

Reply:
Yes, does it fail before this PR?

Reply:
The test (disabling node termination and letting the autoscaler send the drain node request) doesn't make sense before the PR.

Reply:
Isn't the bad log still supposed to be printed before this PR?

Reply:
The fake multinode node provider terminates nodes correctly enough that the useless log doesn't appear, even in master.

In this test, we're at least checking end-to-end that the drain api is correctly called by mocking out the node provider's termination logic.